src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CollectionExpression.cs

```diff
@@ -1070,6 +1070,21 @@ private BoundExpression CreateAndPopulateList(BoundCollectionExpression node, Ty
                         var type = rewrittenSpreadOperand.Type!;
 
                         var useSiteInfo = GetNewCompoundUseSiteInfo();
+
+                        if (spreadElement.EnumeratorInfoOpt is { } enumeratorInfo)
+                        {
+                            var iCollectionOfTType = _compilation.GetSpecialType(SpecialType.System_Collections_Generic_ICollection_T);
+                            var iCollectionOfElementType = iCollectionOfTType.Construct(enumeratorInfo.ElementType);
+
+                            // If collection has a struct enumerator but doesn't implement ICollection<T>
+                            // then manual `foreach` is always more efficient then using `AddRange` method
+                            if (enumeratorInfo.GetEnumeratorInfo.Method.ReturnType.IsValueType &&
+                                !enumeratorInfo.CollectionType.ImplementsInterface(iCollectionOfElementType, ref useSiteInfo))
```

Comment:
Should we maybe use discarded use site info here? Since we only use `ICollection<T>` type for checks, but don't emit it, we don't care about its use site diagnostics, right?